src/daft-plan/src/logical_ops/join.rs

```diff
@@ -70,46 +74,107 @@ impl Join {
+            // old_name -> new_name
+            let right_rename_mapping: HashMap<_, _> = right_names
                 .iter()
-                .map(|(_, field)| field)
-                .cloned()
-                .chain(right.schema().fields.iter().filter_map(|(rname, rfield)| {
-                    if (left_join_keys.contains(rname.as_str())
-                        && right_join_keys.contains(rname.as_str()))
-                        || matches!(join_type, JoinType::Anti | JoinType::Semi)
-                    {
-                        right_input_mapping.insert(rname.clone(), rname.clone());
+                .filter_map(|name| {
+                    if !names_so_far.contains(name) || common_join_keys.contains(name) {
                         None
-                    } else if left_schema.contains_key(rname) {
-                        let new_name = format!("right.{}", rname);
-                        right_input_mapping.insert(new_name.clone(), rname.clone());
-                        Some(rfield.rename(new_name))
                     } else {
-                        right_input_mapping.insert(rname.clone(), rname.clone());
-                        Some(rfield.clone())
+                        let mut new_name = name.clone();
+                        while names_so_far.contains(&new_name) {
+                            new_name = format!("right.{}", new_name);
+                        }
+                        names_so_far.insert(new_name.clone());
+
+                        Some((name.clone(), new_name))
                     }
-                }))
-                .collect::<Vec<_>>();
-            Schema::new(fields).context(CreationSnafu)?.into()
-        };
-        Ok(Self {
-            left,
-            right,
-            left_on,
-            right_on,
-            join_type,
-            join_strategy,
-            output_schema,
-            right_input_mapping,
-        })
+                })
+                .collect();
+
+            let (right, right_on) = if right_rename_mapping.is_empty() {
+                (right, right_on)
+            } else {
+                // projection to update the right side with the new column names
+                let new_right_projection: Vec<_> = right_names
+                    .iter()
+                    .map(|name| {
+                        if let Some(new_name) = right_rename_mapping.get(name) {
+                            Expr::Alias(col(name.clone()), new_name.clone().into()).into()
+                        } else {
+                            col(name.clone())
+                        }
+                    })
+                    .collect();
+
+                let new_right: LogicalPlan =
+                    Project::try_new(right, new_right_projection, Default::default())?.into();
+
+                // change any column references in the right_on expressions to the new column names
+                let new_right_on = right_on
```

Comment:
You can use `replace_columns_with_expressions` here which follows the same logic